### PR TITLE
[Fernflower] autoboxing primitives should not always be ignored

### PR DESCRIPTION
**Pull request by apixandru**
This commit puts more restrictions on when we can safely ignore the boxing primitives.
The initial restrictions were a bit too loose. https://github.com/JetBrains/intellij-community/commit/cdecb09b4a5bdb443d2d7d49698348535ea06b78

### CODE REVIEW

**Comment by apixandru** [comment]
I resolved the merge conflicts.

@gorrus did you get a chance to have a look at this pull request?

**Comment by gorrus** [comment]
Not yet, but it is in the list, sorry for the delay.

File: plugins/java-decompiler/engine/src/org/jetbrains/java/decompiler/modules/decompiler/ExprProcessor.java [hunk 1]
```diff
@@ -540,6 +540,7 @@ else if (cn instanceof LinkConstant) {
             }
 
             InvocationExprent exprinv = new InvocationExprent(instr.opcode, invoke_constant, bootstrap_arguments, stack, bytecode_offsets);
+            markPreviousResultUsedIfBoxingExpression(exprlist, exprinv);
```
Comment (gorrus) [inline plugins/java-decompiler/engine/src/org/jetbrains/java/decompiler/modules/decompiler/ExprProcessor.java]: this looks like an overkill, isn't it better to add something like
`if (instance instanceof InvocationExprent) {
            ((InvocationExprent)instance).markUsingBoxingResult();
          }`
right before `instance.toJava...` inside the InvocationExprent.toJava?

**Comment by apixandru** [comment]
Yes, it does feel like overkill, the new tests pass with your suggested improvement. 

The reason why I had made it so specific was to make sure that this doesn't potentially break any other scenario that I can't think of right now so I made it as specific as I could get.

I removed the highly specific boxing restrictions using another commit rather than amending the previous one just in case it turns out that this is too broad and someone in the future will go through the commit log to see these original restrictions. If you prefer that I squish these commits into one please let me know and I will.

**Comment by gorrus** [comment]
Hi, thanks for changes, please make it in one commit, thanks!

**Comment by apixandru** [comment]
done

**Comment by gorrus** [comment]
Thanks! Manually merged into master.

**Pull request closed without merging**